legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/BaseGitLabApi.java

```diff
@@ -994,4 +1067,540 @@ public String getUserId()
+                    throw new LegendSDLCServerException("Cannot create version " + versionId.toVersionIdString() + " of project " + projectId + ": cannot find current revision (project may be corrupt)", Status.INTERNAL_SERVER_ERROR);
+                }
+            }
+            else
+            {
+                try
+                {
+                    referenceCommit = withRetries(() -> commitsApi.getCommit(projectId.getGitLabId(), revisionId));
+                }
+                catch (GitLabApiException e)
+                {
+                    if (GitLabApiTools.isNotFoundGitLabApiException(e))
+                    {
+                        throw new LegendSDLCServerException("Revision " + revisionId + " is unknown in project " + projectId, Status.BAD_REQUEST);
+                    }
+                    throw e;
+                }
+
+                String sourceBranch = getSourceBranch(projectId, patchReleaseVersion);
+                boolean isOnSourceBranch = PagerTools.stream(withRetries(() -> commitsApi.getCommitRefs(projectId.getGitLabId(), revisionId, CommitRef.RefType.BRANCH, ITEMS_PER_PAGE)))
+                        .anyMatch(ref -> sourceBranch.equals(ref.getName()));
+                if (!isOnSourceBranch)
+                {
+                    throw new LegendSDLCServerException("Revision " + revisionId + " is unknown in project " + projectId, Status.BAD_REQUEST);
+                }
+            }
+
+            Tag tag = gitLabApi.getTagsApi().createTag(projectId.getGitLabId(), tagName, referenceCommit.getId(), message, (String) null);
+            if (notes != null)
+            {
+                gitLabApi.getReleasesApi().createRelease(projectId.getGitLabId(), new ReleaseParams().withTagName(tagName).withDescription(notes));
+            }
+            return fromGitLabTag(projectId.toString(), tag);
+        }
+        catch (Exception e)
+        {
+            throw buildException(e,
+                    () -> "User " + getCurrentUser() + " is not allowed to create version " + versionId.toVersionIdString() + " of project " + projectId,
+                    () -> "Unknown project: " + projectId,
+                    () -> "Error creating version " + versionId.toVersionIdString() + " of project " + projectId);
+        }
+    }
+
+    protected boolean isPatchReleaseBranchPresent(GitLabProjectId projectId, String patchRelaseVersion)
+    {
+        Branch existingBranch = null;
+        try
+        {
+            existingBranch = getGitLabApi().getRepositoryApi().getBranch(projectId.getGitLabId(), getPatchReleaseBranchName(patchRelaseVersion));
+        }
+        catch (Exception e)
+        {
+            LOGGER.warn("Error in fetching the patch release branch", patchRelaseVersion,e);
+        }
+        if (existingBranch == null)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    protected static Constants.MergeRequestState getMergeRequestState(ReviewState state)
```

Comment:
Do we really need to move all these MR/Review methods up to BaseGitLabApi?

Reply:
Last time we discussed we thought we should close all the MRs created for patch branches when they are releases/deleted. To do that i needed all these methods here